### PR TITLE
Debug for issues #50945

### LINKED ISSUE #50945: [e2e test failure] [sig-api-machinery] Aggregator Should be able to support the 1.7 Sample API Server using the current Aggregator
##### Error text:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/aggregator.go:65
creating cluster role wardler
Expected error:
    <*errors.StatusError | 0xc421134380>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {SelfLink: "", ResourceVersion: ""},
            Status: "Failure",
            Message: "clusterroles.rbac.authorization.k8s.io \"wardler\" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:[\"flunders\"], APIGroups:[\"wardle.k8s.io\"], Verbs:[\"create\"]} PolicyRule{Resources:[\"flunders\"], APIGroups:[\"wardle.k8s.io\"], Verbs:[\"delete\"]} PolicyRule{Resources:[\"flunders\"], APIGroups:[\"wardle.k8s.io\"], Verbs:[\"deletecollection\"]} PolicyRule{Resources:[\"flunders\"], APIGroups:[\"wardle.k8s.io\"], Verbs:[\"get\"]} PolicyRule{Resources:[\"flunders\"], APIGroups:[\"wardle.k8s.io\"], Verbs:[\"list\"]} PolicyRule{Resources:[\"flunders\"], APIGroups:[\"wardle.k8s.io\"], Verbs:[\"patch\"]} PolicyRule{Resources:[\"flunders\"], APIGroups:[\"wardle.k8s.io\"], Verbs:[\"update\"]} PolicyRule{Resources:[\"flunders\"], APIGroups:[\"wardle.k8s.io\"], Verbs:[\"watch\"]} PolicyRule{NonResourceURLs:[\"*\"], Verbs:[\"get\"]}] user=&{pr-kubekins@kubernetes-jenkins-pull.iam.gserviceaccount.com  [system:authenticated] map[]} ownerrules=[PolicyRule{Resources:[\"selfsubjectaccessreviews\"], APIGroups:[\"authorization.k8s.io\"], Verbs:[\"create\"]} PolicyRule{NonResourceURLs:[\"/api\" \"/api/*\" \"/apis\" \"/apis/*\" \"/healthz\" \"/swaggerapi\" \"/swaggerapi/*\" \"/version\"], Verbs:[\"get\"]}] ruleResolutionErrors=[]",
            Reason: "Forbidden",
            Details: {
                Name: "wardler",
                Group: "rbac.authorization.k8s.io",
                Kind: "clusterroles",
                UID: "",
                Causes: nil,
                RetryAfterSeconds: 0,
            },
            Code: 403,
        },
    }
    clusterroles.rbac.authorization.k8s.io "wardler" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:["flunders"], APIGroups:["wardle.k8s.io"], Verbs:["create"]} PolicyRule{Resources:["flunders"], APIGroups:["wardle.k8s.io"], Verbs:["delete"]} PolicyRule{Resources:["flunders"], APIGroups:["wardle.k8s.io"], Verbs:["deletecollection"]} PolicyRule{Resources:["flunders"], APIGroups:["wardle.k8s.io"], Verbs:["get"]} PolicyRule{Resources:["flunders"], APIGroups:["wardle.k8s.io"], Verbs:["list"]} PolicyRule{Resources:["flunders"], APIGroups:["wardle.k8s.io"], Verbs:["patch"]} PolicyRule{Resources:["flunders"], APIGroups:["wardle.k8s.io"], Verbs:["update"]} PolicyRule{Resources:["flunders"], APIGroups:["wardle.k8s.io"], Verbs:["watch"]} PolicyRule{NonResourceURLs:["*"], Verbs:["get"]}] user=&{pr-kubekins@kubernetes-jenkins-pull.iam.gserviceaccount.com  [system:authenticated] map[]} ownerrules=[PolicyRule{Resources:["selfsubjectaccessreviews"], APIGroups:["authorization.k8s.io"], Verbs:["create"]} PolicyRule{NonResourceURLs:["/api" "/api/*" "/apis" "/apis/*" "/healthz" "/swaggerapi" "/swaggerapi/*" "/version"], Verbs:["get"]}] ruleResolutionErrors=[]
not to have occurred
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/aggregator.go:331
```
##### Failure cluster statistics:
1 tests failed,    11 jobs failed,    241 builds failed.
Failure stats cover 1 day time range '17 Aug 2017 22:57 UTC' to '18 Aug 2017 22:57 UTC'.
##### Top failed tests by jobs failed:

| Test Name | Jobs Failed |
| --- | --- |
| [sig-api-machinery] Aggregator Should be able to support the 1.7 Sample API Server using the current Aggregator | 11 |

##### Top failed jobs by builds failed:

| Job Name | Builds Failed | Latest Failure |
| --- | --- | --- |
| ci-kubernetes-e2e-gci-gke-multizone | 42 | [18 Aug 2017 22:02 UTC](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gke-multizone/13432) |
| ci-kubernetes-e2e-gci-gke | 40 | [18 Aug 2017 22:00 UTC](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gke/13636) |
| ci-kubernetes-e2e-gke-multizone | 40 | [18 Aug 2017 22:11 UTC](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gke-multizone/14175) |

[Current Status](https://go.k8s.io/triage#42229f8b33f735ea0213)

### PR DESCRIPTION
**Pull request by cheftako**
Aggregator e2e test is intermittantly failing on GKE but not GCE.
Adding the following debugging for help trace issue.
Make sure we always use the same rest client.
Randomly generate the flunder resource name to detect parallel tests.
Print endpoints for sample-system in case multiple instances.
Print original and new pods in case the pod has been restarted.

**What this PR does / why we need it**: Adds debugging for aggregator e2e test to track down GKE flakiness.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #50945 

**Special notes for your reviewer**: This is primarily additional debugging information.

**Release note**:
```release-note NONE
```

### CODE REVIEW

File: test/e2e/apimachinery/aggregator.go [hunk 1]
```diff
@@ -42,6 +43,7 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework"
 	samplev1alpha1 "k8s.io/sample-apiserver/pkg/apis/wardle/v1alpha1"
 
+	"fmt"
```
Comment (lavalamp) [inline test/e2e/apimachinery/aggregator.go]: In the first section is normal for built in packages.

File: test/e2e/apimachinery/aggregator.go [hunk 2]
```diff
@@ -159,7 +161,9 @@ func setupSampleAPIServerCert(namespaceName, serviceName string) *aggregatorCont
 func TestSampleAPIServer(f *framework.Framework, image, namespaceName string) {
 	By("Registering the sample API server.")
 	cleanTest(f, true)
+	rand.Seed(time.Now().UnixNano())
```
Comment (lavalamp) [inline test/e2e/apimachinery/aggregator.go]: Better not to seed the time? I think go seeds by default. You can use the crypto random number package if you are worried.

File: test/e2e/apimachinery/aggregator.go [hunk 3]
```diff
@@ -476,23 +487,43 @@ func TestSampleAPIServer(f *framework.Framework, image, namespaceName string) {
...
 	if len(unstructuredList.Items) != 0 {
 		framework.Failf("failed to get back the correct deleted flunders list %v from the dynamic client", unstructuredList)
 	}
 
 	cleanTest(f, true)
 }
+
+func validateErrorWithDebugInfo(f *framework.Framework, err error, pods *v1.PodList, msg string, fields ...interface{}) {
+	if err != nil {
+		namespace := "sample-system"
```
Comment (lavalamp) [inline test/e2e/apimachinery/aggregator.go]: I thought the framework created a namespace?

**Review by lavalamp — COMMENTED** [review]
A couple thoughts, otherwise lgtm

File: test/e2e/apimachinery/aggregator.go (hunk 1 above)
Comment (cheftako) [reply test/e2e/apimachinery/aggregator.go]: Need to find how to control that in my IDE. Fixed.

File: test/e2e/apimachinery/aggregator.go (hunk 2 above)
Comment (cheftako) [reply test/e2e/apimachinery/aggregator.go]: It does not.
package main

import (
	"fmt"
	"math/rand"
)

func main() {
	fmt.Printf("Hello, playground %d", rand.Int())
}

always generates "Hello, playground 134020434" for me.

File: test/e2e/apimachinery/aggregator.go (hunk 3 above)
Comment (cheftako) [reply test/e2e/apimachinery/aggregator.go]: It does. May have been a bad idea but I wanted to replicate the recommended workflow for a use case like sample-apiserver or service-catalog, which includes creating a namespace. So the test essentially follows those instructions including making its own namespace. Depending on what we find we may remove that step.

File: test/e2e/apimachinery/aggregator.go (hunk 2 above)
Comment (lavalamp) [reply test/e2e/apimachinery/aggregator.go]: Hm. Still, there should be a bunch of calls to rand in our test binaries already, to generate the namespace names if nothing else.

File: test/e2e/apimachinery/aggregator.go (hunk 3 above)
Comment (lavalamp) [reply test/e2e/apimachinery/aggregator.go]: I see--need to make sure this is marked as serial if it does that. Uh, I don't know how one does that nowadays.

File: test/e2e/apimachinery/aggregator.go [hunk 4]
```diff
@@ -159,7 +161,9 @@ func setupSampleAPIServerCert(namespaceName, serviceName string) *aggregatorCont
 func TestSampleAPIServer(f *framework.Framework, image, namespaceName string) {
 	By("Registering the sample API server.")
 	cleanTest(f, true)
```
Comment (liggitt) [inline test/e2e/apimachinery/aggregator.go]: cleanTest deletes the namespace... if you're trying to allow this to run in parallel, you have to pass in what you want cleaned up

File: test/e2e/apimachinery/aggregator.go (hunk 3 above)
Comment (liggitt) [reply test/e2e/apimachinery/aggregator.go]: > the test essentially follows those instructions including making its own namespace

I would just have it use the framework-created namespace

Comment (lavalamp) [reply test/e2e/apimachinery/aggregator.go]: that sounds good, but I'm not going to block this PR on it, since we're trying to get debug signal here.

**Pull request merged**